contrib/cgi/src/cgi.c

```diff
@@ -0,0 +1,683 @@
+static gint64 fread_chunked(char *ptr, size_t size, FILE *stream, GError **error)
+{
+	gint64 items_read = 0;
+	gint64 bytes_read = 0;
+	gint percentage = 0;
+	gint item_size = 8;
+
+	while (bytes_read < (gint64) size) {
+		items_read = fread(ptr + bytes_read, item_size, 1024, stream);
+		bytes_read += items_read * item_size;
+		percentage = (bytes_read * 100) / size;
+		if (!write_upload_status(percentage, error)) {
+			bytes_read = -1;
+			goto out;
+		}
+	}
+
+out:
+	return bytes_read;
+}
+
+/*
+ * Reads stdin and writes to temporary bundle file.
+ */
+static gboolean stdin_to_file(GError **error)
+{
+	gint ret = FALSE;
+	gint64 read_len = 0;
+	gint64 bytes_read = 0;
+	gchar *file_buf = NULL;
+	const gchar *content_length = g_getenv("CONTENT_LENGTH");
+
+	if (!content_length) {
+		g_set_error(error, CGI_ERROR, CGI_ERROR_LENGTH_REQUIRED, "Please provide a Content-Length header.");
+		goto out;
+	}
+
+	if (g_file_test(STATUS_FILE_LOCATION, G_FILE_TEST_EXISTS)) {
+		g_set_error(error, CGI_ERROR, CGI_ERROR_SERVICE_UNAVAILABLE, "Another install operation is currently running (lock file exists).");
+		/* do not clean up the lock file in this case, will reboot shortly! */
+		goto out;
+	}
+
+	/* upload started */
+	write_upload_status(0, error);
+
+	read_len = g_ascii_strtoll(content_length, NULL, 10);
+	if (read_len == 0) {
+		g_set_error(error, CGI_ERROR, CGI_ERROR_BAD_REQUEST, "Content-Length header invalid.");
+		goto error;
+	}
+
+	/* read 'read_len' bytes from stdin ... */
+	file_buf = (char *) g_malloc(read_len);
+	bytes_read = fread_chunked(file_buf, read_len, stdin, error);
+	if (bytes_read == -1 || bytes_read != read_len) {
+		if (!error && !*error)
+			g_set_error(error, CGI_ERROR, CGI_ERROR_BAD_REQUEST, "Content-Length header incorrect.");
+		goto error;
+	}
+
+	if (!g_file_set_contents(BUNDLE_TARGET_LOCATION, file_buf, read_len, error))
+		goto error;
+
+	ret = TRUE;
+
+error:
+	if (*error)
+		/* try to remove status/lock file */
+		g_remove(STATUS_FILE_LOCATION);
+
+out:
+	g_free(file_buf);
+
+	return ret;
+}
+
+/*
+ * Handle CGI env variables and trigger requested operations.
+ */
+static gint cgi_handler(int argc, char **argv)
+{
+	gint ret = 1;
+	RInstaller *installer;
+	const gchar *method = g_getenv("REQUEST_METHOD");
+	const gchar *query_string = g_getenv("QUERY_STRING");
+	GError *error = NULL;
+
+	installer = r_installer_proxy_new_for_bus_sync(G_BUS_TYPE_SYSTEM, G_DBUS_PROXY_FLAGS_GET_INVALIDATED_PROPERTIES, "de.pengutronix.rauc", "/", NULL, &error);
+	if (!installer)
+		goto error;
+
+	if (g_strcmp0("GET", method) == 0) {
+		if (g_strcmp0("progress", query_string) == 0) {
+			/* ?progress shows progress information */
+			if (!progress_status(installer, &error))
+				goto error;
+
+		} else if (g_strcmp0("status", query_string) == 0) {
+			/* ?status shows slot information */
+			if (!slot_status(installer, &error))
+				goto error;
+		} else {
+			g_set_error(&error, CGI_ERROR, CGI_ERROR_BAD_REQUEST, "Resource unavailable");
```

Comment:
I would like a `"Resource '%s' unavailable", query_string` here

Reply:
Good point. I am not sure whether `503 Service Unavailable` is actually the correct HTTP code for this..